### PR TITLE
Expand constraint suggestion related span and add quick fix

### PR DESCRIPTION
**Pull request by weswigham**
This expands the originally very narrow elaboration to be reported any time an unconstrained type parameter is related to something, and adds a quickfix to apply the change.

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@typescript-bot pack this

**Comment by RyanCavanaugh** [comment]
Most of the breaks I saw looked like this, which currently isn't quickfixable. Can we expand it?
```ts
interface Fn<T extends string> {
}

function m<T>(x: Fn<T>) {
}
```

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -18792,8 +18792,13 @@ namespace ts {
...
-                if (strictNullChecks && source.flags & TypeFlags.TypeVariable && source.symbol?.declarations?.[0] && !getConstraintOfType(source as TypeVariable) && isRelatedTo(emptyObjectType, extractTypesOfKind(target, ~TypeFlags.NonPrimitive))) {
-                    associateRelatedInfo(createDiagnosticForNode(source.symbol.declarations[0], Diagnostics.This_type_parameter_probably_needs_an_extends_object_constraint));
+                if (source.flags & TypeFlags.TypeParameter && source.symbol?.declarations?.[0] && !getConstraintOfType(source as TypeVariable)) {
+                    const syntheticparam = cloneTypeParameter(source as TypeParameter);
+                    syntheticparam.constraint = instantiateType(target, makeUnaryTypeMapper(source, syntheticparam));
+                    if (hasNonCircularBaseConstraint(syntheticparam)) {
+                        const targetConstraintString = typeToString(target, source.symbol.declarations[0]);
+                        associateRelatedInfo(createDiagnosticForNode(source.symbol.declarations[0], Diagnostics.This_type_parameter_might_need_an_extends_0_constraint, targetConstraintString));
+                    }
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:18801]:
```suggestion
                    const syntheticParam = cloneTypeParameter(source as TypeParameter);
                    syntheticParam.constraint = instantiateType(target, makeUnaryTypeMapper(source, syntheticParam));
                    if (hasNonCircularBaseConstraint(syntheticParam)) {
                        const targetConstraintString = typeToString(target, source.symbol.declarations[0]);
                        associateRelatedInfo(createDiagnosticForNode(source.symbol.declarations[0], Diagnostics.This_type_parameter_might_need_an_extends_0_constraint, targetConstraintString));
                    }
```

File: src/services/codefixes/fixAddMissingConstraint.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    registerCodeFix({
+        errorCodes,
+        getCodeActions(context) {
+            const { sourceFile, span, program } = context;
+            const related = getDiagnosticRelatedInfo(program, sourceFile, span);
+            if (!related) {
+                return;
+            }
+            const changes = textChanges.ChangeTracker.with(context, t => addMissingConstraint(t, related));
+            return [createCodeFixAction(fixId, changes, Diagnostics.Add_extends_constraint, fixId, Diagnostics.Add_missing_new_operator_to_all_calls)];
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixAddMissingConstraint.ts:24]: > Add_missing_new_operator_to_all_calls

Looks like you accidentally a new diagnostic for the fix-all.

File: src/services/codefixes/fixAddMissingConstraint.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            const info = getDiagnosticRelatedInfo(context.program, context.sourceFile, diag);
+            if (!info) return;
+            return addMissingConstraint(changes, info);
+        }),
+    });
+
+    function getDiagnosticRelatedInfo(program: Program, sourceFile: SourceFile, span: TextSpan) {
+        const diags = program.getSemanticDiagnostics(sourceFile).filter(diag => diag.start === span.start && diag.length === span.length);
+        if (!length(diags)) return;
+        const diag = diags[0];
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixAddMissingConstraint.ts:37]: It sounds like you want `program.getSemanticDiagnostics(sourceFile).find(...)`

Comment (DanielRosenwasser) [reply src/services/codefixes/fixAddMissingConstraint.ts:37]: Somewhat surprised that nobody has had to look at the diagnostics object for a code fix this whole time.

File: src/services/codefixes/fixAddMissingConstraint.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        if (!diag.relatedInformation) return;
+        const related = filter(diag.relatedInformation, related => related.code === Diagnostics.This_type_parameter_might_need_an_extends_0_constraint.code)[0];
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixAddMissingConstraint.ts:39]: Also `.find()` or `ts.forEach(...)`

File: src/services/codefixes/fixAddMissingConstraint.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    function addMissingConstraint(changes: textChanges.ChangeTracker, related: DiagnosticRelatedInformation): void {
+        let decl = findAncestorMatchingSpan(related.file!, related as TextSpan);
+        if (!decl) return;
+        if (isIdentifier(decl) && isTypeParameterDeclaration(decl.parent)) {
+            decl = decl.parent;
+        }
+        if (!isTypeParameterDeclaration(decl) || isMappedTypeNode(decl.parent)) return; // should only issue fix on type parameters written using `extends`
+        const newConstraint = flattenDiagnosticMessageText(related.messageText, "\n", 0).match(/`extends (.*)`/);
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixAddMissingConstraint.ts:51]: Phew, those `` ` ``s are subtle.

Kinda hacky, but the next best thing would be to add some associated info with diagnostics objects too.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by weswigham** [comment]
> Can we expand it?

Sure, should be easy - just another error code in the quickfix hopefully, since it uses a custom head message on the assignability error.

**Comment by weswigham** [comment]
@RyanCavanaugh done. There are a lot of custom head messages for relation errors, and thus, many error codes we could conceivably add to this quickfix. I don't claim the current list to be exhaustive, just the ones I read in the core `reportRelationError` function. 😅

**Comment by RyanCavanaugh** [comment]
@typescript-bot pack this

**Comment by andrewbranch** [comment]
(This isn’t the first time it would have been nice to be able to register a codefix against a _related_ error instead of the head message.)

**Review by andrewbranch — APPROVED** [review]
We should keep an eye out for how often the related diagnostic / quick fix is the correct fix in real world code since it seems to be showing up a lot more often.

**Comment by weswigham** [comment]
Notably, that failure mode _does_ imply another, narrower, potentially useful quickfix - if a _constrained_ type parameter fails being related to some target type, suggesting the target type be intersected into the constraint of the source type parameter is potentially a good fix. Unfortunately, a large corpus of code with errors isn't exactly something we have to know how useful such a fix would be.

**Pull request merged**